src/query/execute_statement.c

```diff
@@ -11802,65 +11797,14 @@ do_insert_template (PARSER_CONTEXT * parser, DB_OTMPL ** otemplate, PT_NODE * st
   /* clear any previous error indicator because the rest of do_insert is sensitive to er_errid(). */
   er_clear ();
 
-  /* fetch the class for instance write purpose */
-  if (!locator_fetch_class (class_->info.name.db_object, DB_FETCH_CLREAD_INSTWRITE))
-    {
-      assert (er_errid () != NO_ERROR);
-      return er_errid ();
-    }
-
-  flag = statement->info.insert.spec->info.spec.flag;
-  if (statement->info.insert.do_replace || statement->info.insert.odku_assignments != NULL)
-    {
-      /* Check to see if the class into which we are inserting is part of an inheritance chain. We do not allow these
-       * statements to be executed in these cases as we might have undefined behavior, such as trying to update a
-       * column that belongs to a child for a duplicate key in the parent table that does not have that column. */
-      int allowed = 0;
-      error = is_replace_or_odku_allowed (class_->info.name.db_object, &allowed);
-      if (error != NO_ERROR)
-	{
-	  goto cleanup;
-	}
-
-      if (!allowed)
-	{
-	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_REPLACE_ODKU_NOT_ALLOWED, 0);
-	  error = er_errid ();
-	  goto cleanup;
-	}
-    }
-
-  error = is_server_insert_allowed (parser, statement);
+  error = do_insert_checks (parser, statement, &class_, &update, value_clauses);
   if (error != NO_ERROR)
     {
+      ASSERT_ERROR ();
       goto cleanup;
     }
 
-  if (statement->info.insert.odku_assignments != NULL)
-    {
-      /* Test if server UPDATE is allowed */
-      update = do_create_odku_stmt (parser, statement);
-      if (update == NULL)
-	{
-	  error = ER_FAILED;
-	  goto cleanup;
-	}
-      if (statement->info.insert.server_allowed == SERVER_INSERT_IS_ALLOWED)
-	{
-	  int server_allowed = 0;
-	  error =
-	    is_server_update_allowed (parser, &statement->info.insert.odku_non_null_attrs, &upd_has_uniques,
-				      &server_allowed, update);
-	  if (error != NO_ERROR)
-	    {
-	      goto cleanup;
-	    }
-	  if (!server_allowed)
-	    {
-	      statement->info.insert.server_allowed = SERVER_INSERT_IS_NOT_ALLOWED;
-	    }
-	}
-    }
+  flag = statement->info.insert.spec->info.spec.flag;
```

Comment:
@lucianlazar2211 

`flag` should be set before a jump to `cleanup`. `info.spec.flag` may be set as 0 without saving the original value.